tdms/src/interpolate.cpp

```diff
@@ -1393,3 +1272,177 @@ void interpolateTimeDomainFieldCentralE_TM(  double ***Exy, double ***Exz, doubl
   *Ey = 0.;
   *Ez = Ezy[k][j][i]+Ezx[k][j][i];
 }
+
+/**
+ * @brief Determines the appropriate interpolation scheme to use
+ * 
+ * @param[in] cells_in_direction The number of Yee cells in the interpolation direction of interest
+ * @param[in] cell_id The current ID (in this dimension) of the Yee cell
+ * @return int Equal to BAND_LIMITED, INTERP1, INTERP2, INTERP3, indicating the appropriate interpolation scheme
+ */
+int determineInterpScheme(int cells_in_direction, int cell_id) {
+  // interpolation is impossible if the total number of cells in this direction is <4
+  if (cells_in_direction <= 4) { throw out_of_range("Error: computational domain has fewer than 4 cells in at least 1 dimension, cubic and band-limited interpolation impossible.\n"); }
+  // Yee cell with index <=0 doesn't exist (indexing starts from 1)
+  else if (cell_id <= 0) { throw out_of_range("Error: Interior Yee cell index <=0 requested (must be >=1).\n");}
+  // Yee cell with index >=cells_in_direction doesn't exist
+  else if (cell_id > cells_in_direction) { throw out_of_range("Error: requested Yee cell index beyond maximum number of Yee cells.\n");}
+
+  // otherwise, now determine which interpolation scheme we should be using
+  int right_cell_buffer = cells_in_direction - cell_id;
+  if ((right_cell_buffer >= 3) && (cell_id >= 4)) {
+    // we have enough room to apply band-limited interpolation
+    return BAND_LIMITED;
+  }
+  else if (right_cell_buffer == 0) {
+    // we are at the extreme right Yee cell, and hence must use interp3
+    return INTERP3;
+  }
+  else if (cell_id == 1) {
+    // we are at the extreme left Yee cell, and must use interp2
+    return INTERP2;
+  }
+  else {
+    // we have sufficient space to use interp2, with two Yee cells either side
+    return INTERP1;
+  }
+}
+
+/**
+ * @brief Interpolate a component of the E-field to the centre of a Yee cell
+ * 
+ * We enforce {a,b,c} = {i,j,k}, with the conventions a=i, b=j, c=k; a=j, b=i, c=k; a=k, b=i, c=j.
+ * Indices and variables are named assuming the a=i case, EG (a,b,c) will specify the index of the Yee cell.
+ * 
+ * @param[in] aID Identifies which direction i(0), j(1), or k(2) the a-direction represents
+ * @param[in] Eab, Eac Split components of the Yee cell
+ * @param[in] a,b,c Index of the Yee cell
+ * @param[in] A Total number of Yee cells in the a-direction
+ * @param[out] Ea The interpolated field component at the centre of the Yee cell
+ */
+void interpolateTimeDomainEcomponent(int aID, double ***Eab, double ***Eac, int a, int b, int c, int A, double *Ea)
```

Comment:
not looked at this yet! need me to understand some more background first 😅

Reply:
You're going to _love_ what I've got so far for the 2D H-field interpolation...